utils/logalloc.cc

```diff
@@ -736,32 +736,36 @@ class seastar_memory_segment_store_backend : public segment_store_backend {
         ::free(seg);
     }
     virtual size_t free_memory() const noexcept override {
-        return memory::stats().free_memory();
+        return memory::free_memory();
     }
 };
 
+static constexpr size_t segment_npos = size_t(-1);
+
 #ifndef SEASTAR_DEFAULT_ALLOCATOR
 class segment_store {
-    memory::memory_layout _layout;
-    uintptr_t _segments_base; // The address of the first segment
+    std::unique_ptr<segment_store_backend> _backend;
 
 public:
     size_t non_lsa_reserve = 0;
     segment_store()
-        : _layout(memory::get_memory_layout())
-        , _segments_base(align_down(_layout.start, (uintptr_t)segment::size)) {
-    }
+        : _backend(std::make_unique<seastar_memory_segment_store_backend>())
+    { }
     const segment* segment_from_idx(size_t idx) const noexcept {
-        return reinterpret_cast<segment*>(_segments_base) + idx;
+        return reinterpret_cast<segment*>(_backend->segments_base()) + idx;
     }
     segment* segment_from_idx(size_t idx) noexcept {
-        return reinterpret_cast<segment*>(_segments_base) + idx;
+        return reinterpret_cast<segment*>(_backend->segments_base()) + idx;
     }
     size_t idx_from_segment(const segment* seg) const noexcept {
-        return seg - reinterpret_cast<segment*>(_segments_base);
+        const auto seg_uint = reinterpret_cast<uintptr_t>(seg);
+        if (seg_uint < _backend->memory_layout().start || seg_uint > _backend->memory_layout().end) [[unlikely]] {
```

Comment:
Shouldn't it be `>=` instead of `>`?

Reply:
End is C++ iterator style, so it is always `<`, or `!=`.

Reply:
Are you sure? By "C++ iterator style", don't you mean that `end` is exclusive?
In that case `seg_uint` is valid when
```seg_uint >= _backend->memory_layout().start && seg_uint < _backend->memory_layout().end``` and invalid when ```seg_uint < _backend->memory_layout().start || seg_uint >= _backend->memory_layout().end``` — the second inequality is non-strict.

Reply:
Ah, you are right, I was thinking of the validity condition (which is `<`), not the invalidity (which is indeed `>=`).